### PR TITLE
switch image.rb to google storage via fog library

### LINKED ISSUE #9760: Implement paperclip or other way to upload to Google Cloud Storage
In order to migrate to Google Cloud Storage, we've had to mount our storage bucket with `gcsfuse`. This solution seems not stable enough as it sometimes locks the mountpoint (#9758 is an instance of this). This happened recently and we didn't notice until today, and it may have caused more problems in the past week. 

**Please describe the desired behavior.**
Ideally, we would not have to mount this volume, but the app would upload the data directly to the Google Cloud Storage API.

### PR DESCRIPTION
**Pull request by jywarren**
Fixes #9760

### CODE REVIEW

File: app/models/image.rb [hunk 1]
```diff
@@ -4,7 +4,10 @@ class Image < ApplicationRecord
   belongs_to :user, foreign_key: :uid
   belongs_to :node, foreign_key: :nid
 
-  has_attached_file :photo, styles: { thumb: '200x150>', medium: '500x375>', large: '800x600>' } # ,
+  has_attached_file :photo, styles: { thumb: '200x150>', medium: '500x375>', large: '800x600>' },
+    :fog_credentials => {  :provider => 'Google',
```
[the remark this answers is not in the document]
Comment (jywarren) [reply app/models/image.rb:8]:
```suggestion
    fog_credentials: {  provider: 'Google',
```

File: app/models/image.rb [hunk 2]
```diff
@@ -4,7 +4,10 @@ class Image < ApplicationRecord
...
+                           :google_storage_access_key_id => ENV["GOOGLE_STORAGE_KEY"],
```
[the remark this answers is not in the document]
Comment (jywarren) [reply app/models/image.rb:9]:
```suggestion
                           google_storage_access_key_id: ENV["GOOGLE_STORAGE_KEY"],
```

File: app/models/image.rb [hunk 3]
```diff
@@ -4,7 +4,10 @@ class Image < ApplicationRecord
...
+                           :google_storage_secret_access_key => ENV["GOOGLE_STORAGE_SECRET"]}
```
[the remark this answers is not in the document]
Comment (jywarren) [reply app/models/image.rb:10]:
```suggestion
                           google_storage_secret_access_key: ENV["GOOGLE_STORAGE_SECRET"]}
```

**Comment by icarito** [comment]
Ok then we'll test this in unstable as discussed!

**Comment by jywarren** [comment]
Tried this out- existing images look good but uploading didn't seem to work:

https://unstable.publiclab.org/notes/keshav_sethi0004/06-07-2020/gsoc-2020-week-5-coding-period-week-1-spam-management-dashboard#c26975

https://storage.googleapis.com/production-public-files/public/system/images/photos/000/039/685/original/PXL_20210610_011822610.jpg

```xml
This XML file does not appear to have any style information associated with it. The document tree is shown below.
<Error>
<Code>NoSuchKey</Code>
<Message>The specified key does not exist.</Message>
<Details>No such object: production-public-files/public/system/images/photos/000/039/685/original/PXL_20210610_011822610.jpg</Details>
</Error>
```

**Comment by jywarren** [comment]
A working image appears at this url:

https://storage.googleapis.com/production-public-files/public/system/images/photos/000/039/683/original/smd.png

**Comment by jywarren** [comment]
We should check the logs, it seems the app is returning a success response so it thinks the file exists and succeeded. The url format looks good. Wonder whats wrong?

**Comment by jywarren** [comment]
Here are the Rails logs for the upload:

```

[7912c979-d527-49fd-b78f-e97ede025ee4] Processing by ImagesController#create as JSON
[7912c979-d527-49fd-b78f-e97ede025ee4]   Parameters: {"uid"=>"1", "image"=>{"photo"=>#<ActionDispatch::Http::UploadedFile:0x00
007f016be18160 @tempfile=#<Tempfile:/tmp/RackMultipart20210617-3597-qxssed.jpg>, @original_filename="PXL_20210610_011822610.jp
g", @content_type="image/jpeg", @headers="Content-Disposition: form-data; name=\"image[photo]\"; filename=\"PXL_20210610_01182
2610.jpg\"\r\nContent-Type: image/jpeg\r\n">}}
[7912c979-d527-49fd-b78f-e97ede025ee4] [paperclip] Trying to link /tmp/RackMultipart20210617-3597-qxssed.jpg to /tmp/2dc06304b
a68b6eb7be46deb4c10885920210617-3597-1vxrxh9.jpg
[7912c979-d527-49fd-b78f-e97ede025ee4] [paperclip] Trying to link /tmp/2dc06304ba68b6eb7be46deb4c10885920210617-3597-1vxrxh9.j
pg to /tmp/2dc06304ba68b6eb7be46deb4c10885920210617-3597-1wefp3t.jpg
[7912c979-d527-49fd-b78f-e97ede025ee4] Command :: file -b --mime '/tmp/2dc06304ba68b6eb7be46deb4c10885920210617-3597-1wefp3t.j
pg'
[7912c979-d527-49fd-b78f-e97ede025ee4] Command :: identify -format '%wx%h,%[exif:orientation]' '/tmp/2dc06304ba68b6eb7be46deb4
c10885920210617-3597-1vxrxh9.jpg[0]' 2>/dev/null
[7912c979-d527-49fd-b78f-e97ede025ee4] Command :: identify -format %m '/tmp/2dc06304ba68b6eb7be46deb4c10885920210617-3597-1vxr
xh9.jpg[0]'
[7912c979-d527-49fd-b78f-e97ede025ee4] Command :: convert '/tmp/2dc06304ba68b6eb7be46deb4c10885920210617-3597-1vxrxh9.jpg[0]' 
-auto-orient -resize "200x150>" '/tmp/473d4bdb9c5b36f63d38026c8de3278220210617-3597-1trlpiq'
[7912c979-d527-49fd-b78f-e97ede025ee4] [paperclip] Trying to link /tmp/473d4bdb9c5b36f63d38026c8de3278220210617-3597-1trlpiq t
o /tmp/c672fcdc5e1ae82e957844125759c1d320210617-3597-1edaa8n
[7912c979-d527-49fd-b78f-e97ede025ee4] Command :: identify -format '%wx%h,%[exif:orientation]' '/tmp/2dc06304ba68b6eb7be46deb4
c10885920210617-3597-1vxrxh9.jpg[0]' 2>/dev/null
[7912c979-d527-49fd-b78f-e97ede025ee4] Command :: identify -format %m '/tmp/2dc06304ba68b6eb7be46deb4c10885920210617-3597-1vxr
xh9.jpg[0]'
[7912c979-d527-49fd-b78f-e97ede025ee4] Command :: convert '/tmp/2dc06304ba68b6eb7be46deb4c10885920210617-3597-1vxrxh9.jpg[0]' 
-auto-orient -resize "500x375>" '/tmp/473d4bdb9c5b36f63d38026c8de3278220210617-3597-ajre7e'
[7912c979-d527-49fd-b78f-e97ede025ee4] [paperclip] Trying to link /tmp/473d4bdb9c5b36f63d38026c8de3278220210617-3597-ajre7e to
 /tmp/ab4648842bc8076ecaeaf0cc7af086ac20210617-3597-1ke7emr
[7912c979-d527-49fd-b78f-e97ede025ee4] Command :: identify -format '%wx%h,%[exif:orientation]' '/tmp/2dc06304ba68b6eb7be46deb4
c10885920210617-3597-1vxrxh9.jpg[0]' 2>/dev/null
[7912c979-d527-49fd-b78f-e97ede025ee4] Command :: identify -format %m '/tmp/2dc06304ba68b6eb7be46deb4c10885920210617-3597-1vxr
xh9.jpg[0]'
[7912c979-d527-49fd-b78f-e97ede025ee4] Command :: convert '/tmp/2dc06304ba68b6eb7be46deb4c10885920210617-3597-1vxrxh9.jpg[0]' 
-auto-orient -resize "800x600>" '/tmp/473d4bdb9c5b36f63d38026c8de3278220210617-3597-1yvqrkp'
[7912c979-d527-49fd-b78f-e97ede025ee4] [paperclip] Trying to link /tmp/473d4bdb9c5b36f63d38026c8de3278220210617-3597-1yvqrkp t
o /tmp/50017c6366a5ce88a96b7dca0d1f333f20210617-3597-1ar3bv8
[7912c979-d527-49fd-b78f-e97ede025ee4] [paperclip] Trying to link /tmp/2dc06304ba68b6eb7be46deb4c10885920210617-3597-1vxrxh9.j
pg to /tmp/2dc06304ba68b6eb7be46deb4c10885920210617-3597-1yy1cq9.jpg
[7912c979-d527-49fd-b78f-e97ede025ee4] Command :: file -b --mime '/tmp/2dc06304ba68b6eb7be46deb4c10885920210617-3597-1yy1cq9.j
pg'
[7912c979-d527-49fd-b78f-e97ede025ee4] Completed 200 OK in 2645ms (Views: 0.4ms | ActiveRecord: 292.9ms)
[85af9bc3-5cf9-4425-b104-cf04f14a1850] Started GET "/i/39685" for 72.92.236.166 at 2021-06-17 11:55:16 +0000
[85af9bc3-5cf9-4425-b104-cf04f14a1850] Processing by ImagesController#shortlink as HTML
[85af9bc3-5cf9-4425-b104-cf04f14a1850]   Parameters: {"id"=>"39685"}
[85af9bc3-5cf9-4425-b104-cf04f14a1850] Redirected to https://unstable.publiclab.org/system/images/photos/000/039/685/original/
PXL_20210610_011822610.jpg
```

**Comment by jywarren** [comment]
OK, strangely, there is already a file in that location, in Google Cloud storage:

https://storage.googleapis.com/production-public-files/public/system/images/photos/000/039/685/original/44.jpg

Created Dec 11 2020, last modified Jan 14, 2021, 12:35:23 PM

So... what's up? 

Wait!!! OMG, it's because unstable is trying to upload into the same production google cloud image store -- we need a different one for the unstable site, i guess?

Or am I just looking in the wrong spot for the file?

**Comment by icarito** [comment]
I found the missing image in the Tycho filesystem.
So I guess this shows that probably the env variables are not being used
and that without them, the system will use the filesystem. I'll review the
env var setup now.
--
Sebastian Silva

**Comment by jywarren** [comment]
I found this log on uploading to unstable:

```
[cf35d0d4-b72f-4826-96d9-fce47858074b] Started POST "/images" for 190.13.155.140 at 2021-06-22 19:24:16 +0000
[cf35d0d4-b72f-4826-96d9-fce47858074b] Processing by ImagesController#create as JSON
[cf35d0d4-b72f-4826-96d9-fce47858074b]   Parameters: {"uid"=>"448589", "image"=>{"photo"=>#<ActionDispatch::Http::UploadedFile:0x00007f8126a38b68 @tempfile=#<Tempfile:/tmp/RackMultipart20210622-3551-1gjscco.jpg>, @original_filename="selfie.jpg", @content_type="image/jpeg", @headers="Content-Disposition: form-data; name=\"image[photo]\"; filename=\"selfie.jpg\"\r\nContent-Type: image/jpeg\r\n">}}
[cf35d0d4-b72f-4826-96d9-fce47858074b] [paperclip] Trying to link /tmp/RackMultipart20210622-3551-1gjscco.jpg to /tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-lr6gpo.jpg
[cf35d0d4-b72f-4826-96d9-fce47858074b] [paperclip] Trying to link /tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-lr6gpo.jpg to /tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-1jhbhin.jpg
[cf35d0d4-b72f-4826-96d9-fce47858074b] Command :: file -b --mime '/tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-1jhbhin.jpg'
[cf35d0d4-b72f-4826-96d9-fce47858074b] Command :: identify -format '%wx%h,%[exif:orientation]' '/tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-lr6gpo.jpg[0]' 2>/dev/null
[cf35d0d4-b72f-4826-96d9-fce47858074b] Command :: identify -format %m '/tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-lr6gpo.jpg[0]'
[cf35d0d4-b72f-4826-96d9-fce47858074b] Command :: convert '/tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-lr6gpo.jpg[0]' -auto-orient -resize "200x150>" '/tmp/20b0c36a9e1d8c5d4f9e9222cac588f520210622-3551-txnmrs'
[cf35d0d4-b72f-4826-96d9-fce47858074b] [paperclip] Trying to link /tmp/20b0c36a9e1d8c5d4f9e9222cac588f520210622-3551-txnmrs to /tmp/b6ae7ad269c0517c1b200a325882455420210622-3551-1rybk5
[cf35d0d4-b72f-4826-96d9-fce47858074b] Command :: identify -format '%wx%h,%[exif:orientation]' '/tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-lr6gpo.jpg[0]' 2>/dev/null
[cf35d0d4-b72f-4826-96d9-fce47858074b] Command :: identify -format %m '/tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-lr6gpo.jpg[0]'
[cf35d0d4-b72f-4826-96d9-fce47858074b] Command :: convert '/tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-lr6gpo.jpg[0]' -auto-orient -resize "500x375>" '/tmp/20b0c36a9e1d8c5d4f9e9222cac588f520210622-3551-w599o5'
[cf35d0d4-b72f-4826-96d9-fce47858074b] [paperclip] Trying to link /tmp/20b0c36a9e1d8c5d4f9e9222cac588f520210622-3551-w599o5 to /tmp/3e79b286047fa0c5214b7fae1ac052a620210622-3551-1ok0dsw
[cf35d0d4-b72f-4826-96d9-fce47858074b] Command :: identify -format '%wx%h,%[exif:orientation]' '/tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-lr6gpo.jpg[0]' 2>/dev/null
[cf35d0d4-b72f-4826-96d9-fce47858074b] Command :: identify -format %m '/tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-lr6gpo.jpg[0]'
[cf35d0d4-b72f-4826-96d9-fce47858074b] Command :: convert '/tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-lr6gpo.jpg[0]' -auto-orient -resize "800x600>" '/tmp/20b0c36a9e1d8c5d4f9e9222cac588f520210622-3551-1mb11ef'
[cf35d0d4-b72f-4826-96d9-fce47858074b] [paperclip] Trying to link /tmp/20b0c36a9e1d8c5d4f9e9222cac588f520210622-3551-1mb11ef to /tmp/6b4b73f4f6abc75d2ce73372f37362ab20210622-3551-jlkp78
[cf35d0d4-b72f-4826-96d9-fce47858074b] [paperclip] Trying to link /tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-lr6gpo.jpg to /tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-1h62l4c.jpg
[cf35d0d4-b72f-4826-96d9-fce47858074b] Command :: file -b --mime '/tmp/f510b87b8c2efb5562be8c33f2bcd30d20210622-3551-1h62l4c.jpg'
[cf35d0d4-b72f-4826-96d9-fce47858074b] Completed 200 OK in 760ms (Views: 0.8ms | ActiveRecord: 209.3ms)
```

**Comment by icarito** [comment]
It doesn't look like an error

> I found this log on uploading to unstable:
>

**Comment by jywarren** [comment]
https://www.rubydoc.info/gems/paperclip/Paperclip/Storage/Fog

https://github.com/fog/fog/pull/2872

**Comment by jywarren** [comment]
```
Progress: |========ERROR ImageTest#test_upload_via_datauri/dataurl (1.43s)
Minitest::UnexpectedError:         NameError: uninitialized constant Fog
            test/unit/image_test.rb:14:in `block in <class:ImageTest>'
```

**Comment by icarito** [comment]
Testing with new HMAC keys, (https://cloud.google.com/storage/docs/authentication/managing-hmackeys#create) after debugging issues with dependencies and configuration.

**Comment by icarito** [comment]
Okay it seems the authentication went fine now, but now we're getting:
```
web_1      | [4e964513-ee0f-472b-8a92-4652f40bffb8] [paperclip] saving images/photos/000/039/705/original/índice.jpg   
web_1      | App 3636 output: [fog][WARNING] Unrecognized arguments: local_root                        
web_1      | [4e964513-ee0f-472b-8a92-4652f40bffb8] Completed 500 Internal Server Error in 1696ms (ActiveRecord: 198.6ms)     
web_1      | [4e964513-ee0f-472b-8a92-4652f40bffb8] Sending event 606c240061b54ef8bb7b2d6eb77f442c to Sentry                  
web_1      | [4e964513-ee0f-472b-8a92-4652f40bffb8]                              
web_1      | [4e964513-ee0f-472b-8a92-4652f40bffb8] Excon::Error::BadRequest (Expected(200) <=> Actual(400 Bad Request)
web_1      | excon.error.response
web_1      |   :body          => "<?xml version='1.0' encoding='UTF-8'?><Error><Code>InvalidArgument</Code><Message>Invalid argument.</Messa
ge><Details>Cannot insert legacy ACL for an object when uniform bucket-level access is enabled. Read more at https://cloud.google.com/storag
e/docs/uniform-bucket-level-access</Details></Error>"                                                                       
web_1      |   :cookies       => [              
web_1      |   ]                                                           
web_1      |   :headers       => {                                        
web_1      |     "Alt-Svc"              => "h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"
; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
web_1      |     "Content-Length"       => "298"                                                                              
web_1      |     "Content-Type"         => "application/xml; charset=UTF-8"
web_1      |     "Date"                 => "Wed, 30 Jun 2021 13:27:56 GMT"
web_1      |     "Server"               => "UploadServer"                        
web_1      |     "Vary"                 => "Origin"
web_1      |     "X-GUploader-UploadID" => "ABg5-UyFXWe4eAIG2MR5i7LHhfYLoSQKQGy5UBiAd62FJ9OZS6oj6vL270ErAfSo6Z03PmrC1IyydIs9kRycewb-YMlJ6-Ey
Jw"
web_1      |   }
web_1      |   :host          => "production-public-files.storage.googleapis.com"
web_1      |   :local_address => "172.19.0.5"
web_1      |   :local_port    => 54676
web_1      |   :path          => "/images%2Fphotos%2F000%2F039%2F705%2Foriginal%2F%C3%ADndice.jpg"
web_1      |   :port          => 443
web_1      |   :reason_phrase => "Bad Request"
web_1      |   :remote_ip     => "142.250.80.16"
web_1      |   :status        => 400
web_1      |   :status_line   => "HTTP/1.1 400 Bad Request\r\n"
web_1      | ):
```

**Comment by icarito** [comment]
so it looks like the `local_root` variables is required when using `fog-local`  but rejected when using `fog-google`

**Comment by icarito** [comment]
Looks like the problem wasn't local_root, will restore it for now and figure out the actual problem.
```
web_1      | [82e9879b-4558-4d97-b359-4306843c2902] [paperclip] saving images/photos/000/039/706/original/publiclab.png            [70/1934]
web_1      | [82e9879b-4558-4d97-b359-4306843c2902] Completed 500 Internal Server Error in 2169ms (ActiveRecord: 612.6ms)                   
web_1      | [82e9879b-4558-4d97-b359-4306843c2902] Sending event 68d1e4c30ad145b385ca290b8dce2dd7 to Sentry                                
web_1      | [82e9879b-4558-4d97-b359-4306843c2902]                                                                                         
web_1      | [82e9879b-4558-4d97-b359-4306843c2902] Excon::Error::BadRequest (Expected(200) <=> Actual(400 Bad Request)                     
web_1      | excon.error.response                                                                                                           
web_1      |   :body          => "<?xml version='1.0' encoding='UTF-8'?><Error><Code>InvalidArgument</Code><Message>Invalid argument.</Mess$
ge><Details>Cannot insert legacy ACL for an object when uniform bucket-level access is enabled. Read more at https://cloud.google.com/stora$
e/docs/uniform-bucket-level-access</Details></Error>"                                                                                       
web_1      |   :cookies       => [                                                                          
web_1      |   ]                                                                                                                            
web_1      |   :headers       => {                                                                    
web_1      |     "Alt-Svc"              => "h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\$
; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""                              
web_1      |     "Content-Length"       => "298"                                           
web_1      |     "Content-Type"         => "application/xml; charset=UTF-8"                                         
web_1      |     "Date"                 => "Wed, 30 Jun 2021 14:44:03 GMT"                                                                  
web_1      |     "Server"               => "UploadServer"                                                   
web_1      |     "Vary"                 => "Origin"                                                    
web_1      |     "X-GUploader-UploadID" => "ABg5-UzYoFFjaa5iSmawrabiqbJb0VUg4L5J2BfEgKHEWiry13kAaZ_J59gU5OgRrdHgh5tQ5J46qDjAMt1wDYfiiD8"
web_1      |   }                                                                                        
web_1      |   :host          => "production-public-files.storage.googleapis.com"                     
web_1      |   :local_address => "172.21.0.5"                                                 
web_1      |   :local_port    => 52138                                                            
web_1      |   :path          => "/images%2Fphotos%2F000%2F039%2F706%2Foriginal%2Fpubliclab.png"  
web_1      |   :port          => 443                                                                  
web_1      |   :reason_phrase => "Bad Request"                                                                       
web_1      |   :remote_ip     => "142.250.80.16"                                                           
web_1      |   :status        => 400                                                                 
web_1      |   :status_line   => "HTTP/1.1 400 Bad Request\r\n"                                               
web_1      | ):                                                                                      
web_1      | [82e9879b-4558-4d97-b359-4306843c2902]                                                
web_1      | [82e9879b-4558-4d97-b359-4306843c2902] app/controllers/images_controller.rb:37:in `create' 
```

**Comment by icarito** [comment]
The docs say:
```
There are two ways to access Google Cloud Storage. The old S3 API and the new JSON API. Fog::Storage::Google will automatically direct you to the appropriate API based on the credentials you provide it.

    The XML API is almost identical to S3. Use Google's interoperability keys to access it.
    The new JSON API is faster and uses auth similarly to the rest of the Google Cloud APIs using a service account private key.
```
So I'm guessing the ACL error that happens with `uniform bucket level access` happens when using the XML S3 API... Will try the JSON one.

File: Gemfile [hunk 4]
```diff
@@ -12,6 +12,8 @@ gem 'authlogic-oid'
 gem "composite_primary_keys"
 gem 'figaro' # To easily test OAuth providers in the development mode
 gem 'friendly_id'
+gem 'fog-google', '1.13.0'
```
Comment (jywarren) [inline Gemfile:15]: Limited to `v1.13.0` due to https://github.com/fog/fog-google/issues/535
```suggestion
gem 'fog-google', '1.13.0' # Limited to `v1.13.0` due to https://github.com/fog/fog-google/issues/535
```

**Comment by jywarren** [comment]
I opened an issue at https://github.com/fog/fog-google/issues/537 to figure out how to get past this error:

```
Google::Apis::ClientError

invalid: Cannot insert legacy ACL for an object when uniform bucket-level access is enabled. Read more at https://cloud.google.com/storage/docs/uniform-bucket-level-access
```

https://sentry.io/share/issue/393f9e2786b543be9b2061a933268129/

**Comment by jywarren** [comment]
Next step - if we don't hear from https://github.com/fog/fog-google/issues/537 by Tuesday we'll clone the bucket without uniform bucket-level access.

Alt plan: look at disabling ACL features on Paperclip?

File: config/initializers/paperclip.rb [hunk 5]
```diff
@@ -8,3 +8,15 @@
...
 }
 Paperclip::DataUriAdapter.register
+Paperclip::Attachment.default_options[:storage] = :fog
+Paperclip::Attachment.default_options[:fog_directory] = ENV["GOOGLE_STORAGE_BUCKET_NAME"] || ''
+Paperclip::Attachment.default_options[:path] = ":rails_root/public/system/system/:class/:attachment/:id_partition/:style/:filename"
+Paperclip::Attachment.default_options[:fog_credentials] = {
+    provider: ENV["FOG_PROVIDER"] || "Local",
+    local_root: "#{Rails.root}/public",
+    google_project: 'public-lab' ,
+    google_json_key_location: ENV["GOOGLE_JSON_KEY_FILE"] || '' ,
```
Comment (jywarren) [inline config/initializers/paperclip.rb]: Removed the XML keys below this line after input from folks at https://github.com/fog/fog-google/issues/537#issuecomment-877544092

```ruby
    google_storage_access_key_id: ENV["GOOGLE_STORAGE_KEY"] || '' ,
    google_storage_secret_access_key: ENV["GOOGLE_STORAGE_SECRET"] || ''
```

**Comment by ebarry** [comment]
Thanks for working on this!

**Comment by icarito** [comment]
We've been looking at changing the `:url` default settings for paperclip but it seems it doesn't have any effect, maybe an instance of https://github.com/spree/spree/issues/4448#issuecomment-37734007

**Comment by icarito** [comment]
I've come up with a workaround for the wrong path issue, will test it out in GitPod.
It involves adding another redirection in Nginx configuration for paths `public/system/images` in addition to the existing one for `system/images`

**Comment by icarito** [comment]
images aren't visible in gitpod with current workaround, will look at solutions

**Comment by jywarren** [comment]
Hi, i just tested this in GitPod - and it worked, for new images:

https://3000-jade-turkey-l45fy0gq.ws-us11.gitpod.io/i/1.png was the upload, which led to 
https://3000-jade-turkey-l45fy0gq.ws-us11.gitpod.io/public/system/images/photos/000/000/001/original/aseprite_templates.png

Is it old images which aren't working? I'm actually not sure how to test old ones on GitPod, so i didn't think that's what you meant. Does this suggest maybe we're good to go here?

**Comment by icarito** [comment]
It's old images for instance from the dashboard that don't show...

**Comment by jywarren** [comment]
I think we're ready to go! The gitpod dev setup has no old images. Let's pass tests and merge, final check stable, and publish!

**Comment by jywarren** [comment]
Woohoo!! 🎉

**Pull request merged**